tests/forms_tests/field_tests/test_urlfield.py

```diff
@@ -7,150 +7,117 @@
+            ),
+            # Normalization.
+            ('http://example.com/     ', 'http://example.com/'),
+            # Valid IDN.
+            ('http://עברית.idn.icann.org/', 'http://עברית.idn.icann.org/'),
+            ('http://sãopaulo.com/', 'http://sãopaulo.com/'),
+            ('http://sãopaulo.com.br/', 'http://sãopaulo.com.br/'),
+            ('http://пример.испытание/', 'http://пример.испытание/'),
+            ('http://مثال.إختبار/', 'http://مثال.إختبار/'),
+            ('http://例子.测试/', 'http://例子.测试/'),
+            ('http://例子.測試/', 'http://例子.測試/'),
+            ('http://उदाहरण.परीक्षा/', 'http://उदाहरण.परीक्षा/',),
+            ('http://例え.テスト/', 'http://例え.テスト/'),
+            ('http://مثال.آزمایشی/', 'http://مثال.آزمایشی/'),
+            ('http://실례.테스트/', 'http://실례.테스트/'),
+            ('http://العربية.idn.icann.org/', 'http://العربية.idn.icann.org/'),
+            # IPv6.
+            ('http://[12:34::3a53]/', 'http://[12:34::3a53]/'),
+            ('http://[a34:9238::]:8080/', 'http://[a34:9238::]:8080/'),
+        ]
+        for url, expected in tests:
             with self.subTest(url=url):
-                # Valid IDN
-                self.assertEqual(url, f.clean(url))
+                self.assertEqual(f.clean(url), expected)
 
-    def test_urlfield_10(self):
-        """URLField correctly validates IPv6 (#18779)."""
+    def test_urlfield_clean_invalid(self):
         f = URLField()
-        urls = (
-            'http://[12:34::3a53]/',
-            'http://[a34:9238::]:8080/',
-        )
-        for url in urls:
-            with self.subTest(url=url):
-                self.assertEqual(url, f.clean(url))
+        tests = [
+            'foo',
+            'com.',
+            '.',
+            'http://',
+            'http://example',
+            'http://example.',
+            'http://.com',
+            'http://invalid-.com',
+            'http://-invalid.com',
+            'http://inv-.alid-.com',
+            'http://inv-.-alid.com',
+            '[a',
+            'http://[a',
+            # Non-string.
+            23,
+            # Hangs "forever" before fixing a catastrophic backtracking,
+            # see #11198.
+            'http://%s' % ('X' * 60,),
+            # A second example, to make sure the problem is really addressed,
+            # even on domains that don't fail the domain label length check in
+            # the regex.
+            'http://%s' % ("X" * 200,),
+        ]
+        msg = "'Enter a valid URL.'"
+        for value in tests:
+            with self.subTest(value=value):
+                with self.assertRaisesMessage(ValidationError, msg):
+                    f.clean(value)
+
+    def test_urlfield_clean_required(self):
+        f = URLField()
+        msg = "'This field is required.'"
+        with self.assertRaisesMessage(ValidationError, msg):
+            f.clean(None)
+        with self.assertRaisesMessage(ValidationError, msg):
+            f.clean('')
```

Comment:
Can't we use `subTest` in this test?

Reply:
It seems like overkill just for two cases, and it's unlikely that we will have more.